### PR TITLE
[inductor] Fix FlexAttention shared memory OOM on consumer GPUs

### LINKED ISSUE #133254: Shared memory out of resource when using flex attention
### 🐛 Describe the bug

When I use flex attention on one RTX 4090, I got some error.
A minimal repro:
```python
import torch
from torch.nn.attention.flex_attention import flex_attention
flex_attention = torch.compile(flex_attention, dynamic=False)

B = 1
H = 16
S = 2048
D = 128

query = torch.randn(B, H, S, D, device="cuda", dtype=torch.bfloat16, requires_grad=True)
key = torch.randn(B, H, S, D, device="cuda", dtype=torch.bfloat16, requires_grad=True)
value = torch.randn(B, H, S, D, device="cuda", dtype=torch.bfloat16, requires_grad=True)
gradOut = torch.randn(B, H, S, D, device="cuda", dtype=torch.bfloat16)

flex_out = flex_attention(query, key, value)
flex_out.backward(gradOut, retain_graph=True)
```
It works with D=64 or without compile.
(I see https://github.com/pytorch/pytorch/issues/132075, while this happens during backward)

### Error logs

```
triton.runtime.errors.OutOfResources: out of resource: shared memory, Required: 131074, Hardware limit: 101376. Reducing block sizes or `num_stages` may help.
```

### Versions

torch==2.5.0.dev20240810+cu121

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @drisspg 
```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by SakshamKapoor2911**
Add fallback configs and shared memory estimation to FlexAttention to prevent OutOfResources errors on low-shared-memory CUDA devices (RTX 4090/3090/A10G).

The approach follows Jason Ansel's PR #184310 with improvements:
- Normalized forward/backward shared-memory overhead constants (both use 4096;
  backward was 8 in #184310, leading to underestimated safety margins)
- Added documentation for the overhead constants

Before (non-autotune mode, D>=128):
```
triton.runtime.errors.OutOfResources: out of resource: shared memory,
Required: 131074, Hardware limit: 101376
```

Fix:
- Estimate forward/backward kernel shared memory from block sizes and head dims
- Filter configs that exceed device shared memory before compilation
- Generate smaller fallback configs (reduced block sizes, fewer stages) when not in max_autotune mode
- Early-exit on first successful config in non-autotune mode

closes: #133254

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by SakshamKapoor2911** [comment]
This is a rework of Jason Ansel's #184310 (stalled for 2+ months), adapted to current main with the overhead constants normalized. See the discussion there for the original approach.

**Comment by jansel** [comment]
@claude review these changes

**Comment by SakshamKapoor2911** [comment]
@jansel -- thanks for the thorough review, really helpful catches. Re-pushed with fixes for both blocking issues and the main design concerns.

**Blocking fixes:**
- Replaced all 6 `len(configs)==1` sites with `not config.max_autotune`. Tile shrinking, divisibility errors, and template error propagation now fire correctly in non-autotune mode regardless of config list length. `test_invalid_block_size` should raise again.
- Fixed forward test call signature (missing `seq_len` argument).

**Design:**
- Switched from `triton.compiler.compiler.max_shared_mem` (Triton internal) to `torch.cuda.get_device_properties` (stable API, supports ROCm `shared_memory_per_block` too).
- Added `log.info` when configs are pruned by shared memory, so silent perf drops are visible.
- Added explicit `RuntimeError` when *all* configs are filtered by shared memory (both fwd and bwd). No more `NoValidChoicesError`.
- Documented the 4096 overhead constants with measurement context (RTX 3090 / sm_86, Triton 3.2.x, non-TMA, non-warp-spec).
- Added comment justifying why forward estimate omits `num_stages` (single-stage kernel vs two-pass backward).

**Intentional divergences from the review:**
- Did not gate pruning behind `max_autotune_prune_choices_based_on_shared_mem` -- that flag is off by default and meant for GEMM pruning where estimates over-prune. Gating here would mean the OOM fix does not work for default users. The `log.info` covers the observability concern.
- Kept smem filtering in the lowering instead of the heuristics layer -- the lowering already has the kernel options dict fully populated (QK_HEAD_DIM_ROUNDED, V_HEAD_DIM_ROUNDED via `set_head_dim_values`). Plumbing that through the heuristics API would require more refactoring than justified.

**Deferred (not yet addressed -- happy to do these before landing if you prefer):**
- Remove the `_exceeds_max_shared_memory` `Callable` indirection
- Fix `_append_unique_config` type erasure (`list[Any]`)
- Add stronger test assertions (verify smaller BLOCK was selected, not just correct output shape)
- Convert golden asserts to property-based
- Include FP32 accumulator in forward estimate (accumulators are register-based in Triton, so likely captured by the overhead constant already)

Ready for another look.

**Comment by SakshamKapoor2911** [comment]
Follow-up on the deferred items from my last review — all now addressed in the latest push:

- **Removed the `_exceeds_max_shared_memory` Callable indirection** — the comparison is now inlined at both the forward and backward call sites.
- **Fixed `_append_unique_config` type erasure** — now typed as `list[_T]` / `list[_T]` instead of `list[Any]`.
- **Stronger test assertions** — the new tests now assert the smaller block was actually selected via `FileCheck` on `BLOCK_M`/`BLOCK_N` in the generated kernel, not just output shape.
- **Golden asserts → property-based** — replaced `assertEqual(configs[0], ...)` with `assertIn(...)` for a strictly-smaller fallback config plus `_assert_fallback_configs_smaller_than_default`, which checks every fallback is smaller than the default on at least one block attribute.
- **FP32 accumulator in the forward estimate** — intentionally omitted, with an added comment explaining Triton keeps accumulators in registers (not shared memory), and register pressure is already reflected in occupancy constraints.

The two intentional divergences (no `max_autotune_prune_choices_based_on_shared_mem` gate; smem filtering kept in the lowering) are unchanged.

Also fixed the PYFMT lint failure in `test_flex_attention.py` that CI flagged; the full non-clang lint set passes locally now.

**Comment by SakshamKapoor2911** [comment]
@drisspg — the `inductor` workflow has 54 checks green; the remaining `pull`, `Lint`, `docs-build`, and `BC Lint` workflows still need approval (scroll to bottom of the PR page). All 6 `len(configs)==1` sites have been migrated to `not config.max_autotune`, the deferred items are addressed, and local tests pass (4/4 CUDA including `test_invalid_block_size`). Could you approve the remaining workflows? @jansel for visibility.

**Comment by jansel** [comment]
@drisspg should review this one

**Comment by jansel** [comment]
@drisspg ping again

**Review by drisspg — CHANGES_REQUESTED** [review]
Hey sorry for no reply in a while
I had two alternative methods for solving;

OG; https://github.com/pytorch/pytorch/pull/172978
part 2; https://github.com/pytorch/pytorch/pull/186880


my preference is for the second which is not what this PR does. I have foudn the auto fallback to be magical and there are 2 quite well documented ways that ultimately get you a better result; which 
1. enable max-autotune
2. manually sweep the kernel options for fwd and backward for your particular score/maskmod

**Comment by SakshamKapoor2911** [comment]
﻿@drisspg Understood. I'll go ahead and close this PR in favor of #186880. Thanks for the context and review!
